### PR TITLE
Multiple Improvements

### PR DESCRIPTION
**Pull request by korydraughn**
Major updates:
- Ticket support for various operations
- Merging /metadata into various endpoints
- Atomic ACL modifications for /collections and /data-objects
- Connection pool refresh configuration options
- etc.

Something I noticed while writing the test for detecting resource changes was using seconds for the mtime does not provide enough precision for the connection pool to detect changes in all use-cases (e.g. very fast resource updates via a test).

Another thing that occurred to me while working on this was how to deal with the connection pool having a later mtime for a resource which no longer exists. I haven't determined whether we need to add logic for this case yet. If we do, that will be a change to the connection pool implementation in the irods/irods repo.

As of right now, this PR passes all tests consistently. I'm going to do another pass over it as there are a few more things I want to do before merging it.

I may wait until #75 is merged before merging this PR.

If interested, feel free to look through the PR and leave comments.

### CODE REVIEW

**Comment by trel** [comment]
> how to deal with the connection pool having a later mtime for a resource which no longer exists.

oh, very interesting.     could... return the count of resources as well as the latest timestamp...  if timestamp goes up OR count goes down... then refresh.

File: CMakeLists.txt
Comment (trel) [inline CMakeLists.txt:76]: how does this commented line align with the initial comment above?

File: README.md
Comment (trel) [inline README.md:175]: NFSRODS uses... 
```
    "proxy_admin_account": {
```

File: src/endpoints/CMakeLists.txt
Comment (trel) [inline src/endpoints/CMakeLists.txt:24]: is this coming soon?   or should be removed?

File: src/endpoints/authentication/CMakeLists.txt
Comment (trel) [inline src/endpoints/authentication/CMakeLists.txt:22]: same

File: src/endpoints/data_objects/impl.cpp
Comment (trel) [inline src/endpoints/data_objects/impl.cpp:337]: this block of code is repeated... looks like 7-8x so far... any sane way to unify?

**Comment by korydraughn** [comment]
Right. I'll capture it in an issue and get a fix in for the connection pool.

What are your thoughts on the mtime precision for resource updates? How do you feel about changing the modify_ts column to hold milliseconds? Or perhaps you feel adding a new column is more appropriate (not sure I like this even though it's safe).

I think all timestamp columns need to be changed to use milliseconds rather than seconds.

File: src/endpoints/data_objects/impl.cpp
Comment (trel) [inline src/endpoints/data_objects/impl.cpp:540]: do we need to consider hierarchies here?     probably already handled... but just leaving a note.

File: test.sh
Comment (trel) [inline test.sh:512]: let's remove kory from the test

File: test/test_irods_http_api.py
Comment (trel) [inline test/test_irods_http_api.py:675]: space between 'A' and *

**Review by trel — COMMENTED** [review]
saw a number of assertions commented out in the tests.

this is very impressive.

File: CMakeLists.txt
Comment (korydraughn) [reply CMakeLists.txt:76]: That is one of the things I'm planning on revisiting before merging this. So, the short answer is, I haven't spent the time thinking about it yet. I also haven't confirmed this works. Will do that soon though.

**Comment by trel** [comment]
> I think all timestamp columns need to be changed to use milliseconds rather than seconds.

That's a really big change.   Many systems didn't support/report greater than 1-second precision for a long time.   So seconds was what was stored, and stored as epoch seconds because timezones are really hard.

We could look into it, but it would have to be a .0 type of thing, and maybe even 5.0.

I think we run with 1-second precision for now for this feature.

File: README.md
Comment (korydraughn) [reply README.md:175]: Right. I made this change before I realized that.

Do we want to use the NFSRODS name?

File: CMakeLists.txt
Comment (trel) [reply CMakeLists.txt:76]: Gotcha.   Resolve whenever you feel ready.

File: README.md
Comment (trel) [reply README.md:175]: yes please - it still looks good and is clean/clear.

File: src/endpoints/CMakeLists.txt
Comment (korydraughn) [reply src/endpoints/CMakeLists.txt:24]: I think that is what we want, but it doesn't work for some reason. I'm going to look into it to try and learn why before I reach out to @SwooshyCueb.

File: src/endpoints/data_objects/impl.cpp
Comment (korydraughn) [reply src/endpoints/data_objects/impl.cpp:337]: Yes, there is. I'm going to take care of that soon.

File: src/endpoints/data_objects/impl.cpp
Comment (korydraughn) [reply src/endpoints/data_objects/impl.cpp:540]: So far, none of the operations support hierarchies. That's not to say they won't, no thought has been spent on supporting them yet.

Willing to discuss whether they should be supported though.

File: src/endpoints/data_objects/impl.cpp
Comment (trel) [reply src/endpoints/data_objects/impl.cpp:337]: ah, very nice.

File: test.sh
Comment (korydraughn) [reply test.sh:512]: This file may go away entirely since the official test suite is written in python.

File: src/endpoints/data_objects/impl.cpp
Comment (trel) [reply src/endpoints/data_objects/impl.cpp:540]: no interest in supporting them.   just want to make sure this code works as desired when handed one.   aka, does it error out correctly/as expected?

**Comment by korydraughn** [comment]
Switching to millisecond precision is a no go due to the values becoming too large to represent over time. However, we can introduce a new column that holds the fractional part of the mtime.

The new column could be in milliseconds or even microseconds and initialized to 0 making it completely safe.

File: src/endpoints/CMakeLists.txt
Comment (SwooshyCueb) [reply src/endpoints/CMakeLists.txt:24]: We don't want to put targets in `target_include_directories` at all. They go in `target_link_libraries`.

**Review by SwooshyCueb — CHANGES_REQUESTED** [review]

File: src/endpoints/CMakeLists.txt
Comment (korydraughn) [reply src/endpoints/CMakeLists.txt:24]: The commented out code certainly didn't work. Things only compiled when using `${IRODS_EXTERNALS_FULLPATH_JSON}/include`.

I'm guessing we aren't doing what CURL is doing then? I feel like I've seen `nlohmann_json::nlohmann_json` in an include directive before.

Comment (SwooshyCueb) [reply src/endpoints/CMakeLists.txt:24]: `CURL::libcurl` should be in `target_link_libraries`, too. Passing a target to `target_link_libraries` takes care of includes as well as any library linkages. You shouldn't need to *ever* reference `IRODS_EXTERNALS_FULLPATH_JSON` unless something has gone horribly wrong.

Comment (korydraughn) [reply src/endpoints/CMakeLists.txt:24]: Oh. I had no idea `target_link_libraries` handled include paths. I'll remove that and see what happens.

File: src/endpoints/data_objects/impl.cpp
Comment (korydraughn) [reply src/endpoints/data_objects/impl.cpp:540]: I'd expect either an error or undefined behavior in the context of iRODS.

Also, dstream doesn't support hierarchies.

Comment (trel) [reply src/endpoints/data_objects/impl.cpp:540]: fair enough - resolving.

File: src/endpoints/data_objects/impl.cpp
Comment (korydraughn) [reply src/endpoints/data_objects/impl.cpp:337]: The more I look at these, the more I feel it is fine.

The only lines that seem worthy of unifying are the ones dealing with constructing an error response.

If we try and encapsulate the inner most if-statement, we either have to throw an exception which isn't ideal (since we can just return) or we have to return more info from the new function.

I'm going to consolidate the body of the inner if-statement.

Comment (korydraughn) [reply src/endpoints/data_objects/impl.cpp:337]: Created issue #85 for this.

Leaving as is for now.

File: README.md
Comment (korydraughn) [reply README.md]: Done.

File: src/endpoints/CMakeLists.txt
Comment (korydraughn) [reply src/endpoints/CMakeLists.txt:24]: Done.

File: src/endpoints/authentication/CMakeLists.txt
Comment (korydraughn) [reply src/endpoints/authentication/CMakeLists.txt]: Done.

File: test.sh
Comment (korydraughn) [reply test.sh:523]: Created issue #86 to address this.

File: test/test_irods_http_api.py
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done.

File: src/endpoints/query/impl.cpp
Comment (korydraughn) [inline src/endpoints/query/impl.cpp:178]: Should the query endpoint accept tickets?

`iquest` doesn't expose a way to enable tickets, but it could. Perhaps there's a reason behind it not exposing a ticket option.

File: src/endpoints/shared_api_operations.cpp
Comment (korydraughn) [inline src/endpoints/shared_api_operations.cpp:11]: Remove empty line.

File: src/endpoints/query/impl.cpp
Comment (korydraughn) [reply src/endpoints/query/impl.cpp:178]: After some discussion, we've concluded the /query endpoint should not accept tickets.

Only the following operations should accept tickets.
- Stat'ing a data object
- Stat'ing a collection
- Writing to a data object
- Reading a data object

Comment (korydraughn) [reply src/endpoints/query/impl.cpp]: Done

File: src/endpoints/shared_api_operations.cpp
Comment (korydraughn) [reply src/endpoints/shared_api_operations.cpp]: Done

File: CMakeLists.txt
Comment (korydraughn) [reply CMakeLists.txt]: Done

**Comment by korydraughn** [comment]
All comments have been addressed.

All tests pass with ASan enabled!

Squashing.

**Comment by korydraughn** [comment]
This PR is in a good spot for merging.

Taking out of draft mode.

**Comment by trel** [comment]
How much does it conflict with #75?     Let's merge ASAP if you're good with it.   Fewer dangling pieces out there.

**Comment by korydraughn** [comment]
There shouldn't be too much conflict since #75 is dealing with auth stuff. This PR touches the auth endpoint, but only for trivial changes.

**Comment by trel** [comment]
Agreed - if it rebases cleanly in a test branch, we'll know for sure.

**Comment by korydraughn** [comment]
This PR touches the auth endpoint in the following ways:
- Fixes a memory leak
- Renames the `rodsadmin` config property to `proxy_admin_account`
- Tweaks the CMakeLists.txt file

**Comment by trel** [comment]
Okay, easy enough to see then...

Yes, please merge this one when #'d and ready.

**Comment by korydraughn** [comment]
Added #'s

**Pull request merged**
